lib/src/annotate.rs

```diff
@@ -0,0 +1,335 @@
+// Copyright 2020 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Methods that allow annotation (attribution and blame) for a file in a
+//! repository
+
+use std::collections::HashMap;
+use std::io::BufRead;
+
+use thiserror::Error;
+
+use crate::backend::{BackendError, BackendResult, CommitId, FileId, TreeValue};
+use crate::commit::Commit;
+use crate::diff::{Diff, DiffHunk};
+use crate::merged_tree::MergedTree;
+use crate::object_id::ObjectId;
+use crate::repo::{ReadonlyRepo, Repo};
+use crate::repo_path::{RepoPath, RepoPathBuf};
+use crate::store::Store;
+
+/// Various errors that can arise from annotation
+#[derive(Debug, Error)]
+pub enum AnnotateError {
+    /// the requested file path was not found
+    #[error("Unable to locate file: {0}")]
+    FileNotFound(String),
+    /// the file type is incorrect. Usually a directory was given but a regular
+    /// file is required
+    #[error("File {0} must be a regular file, not a directory")]
+    UnsupportedFileType(String),
+    /// the file is in a conflicted state and can therefore not be annotated
+    /// properly
+    #[error("File {0} is conflicted at commit: {0}")]
+    Conflicted(String, String),
+    /// pass-through of uncaught backend errors
+    #[error("{0}")]
+    BackendError(BackendError),
+}
+
+impl From<BackendError> for AnnotateError {
+    fn from(value: BackendError) -> Self {
+        AnnotateError::BackendError(value)
+    }
+}
+
+/// Annotation results for a specific file
+pub struct AnnotateResults {
+    /// An array of annotation results ordered by line.
+    /// For each value in the array, the commit_id is the commit id of the
+    /// originator of the line and the string is the actual line itself (without
+    /// newline terminators). The vector is ordered by appearance in the
+    /// file
+    pub file_annotations: Vec<(CommitId, String)>,
+}
+
+struct PartialResults {
+    original_line_map: HashMap<usize, CommitId>,
+    local_line_map: HashMap<CommitId, HashMap<usize, usize>>, /* CommitId -> (local_line_index
```

Comment:
Nit: missing documentation.

Reply:
Taking notes: `original_line_map` seems to map the line number of a line in the file in the original commit to its annotation. That is what's used to make "AnnotateResults".

I'm guessing that `local_line_map` only tracks those lines in the given commit that correspond to some line in the original commit.

------

IIUC, the CommitId in `original_line_map` is just an annotation, whereas the CommitId in `local_line_map` tracks the progress of the algorithm, and for *those* CommitId-s it's important that you can extract the parent commit(s) from them. It might help if you do `type Annotation = CommitId` and replace the CommitId-s that are just annotations with them. For one, annotations will hopefully include a conflict side id at some point.

Similarly, `type LineNumber=usize` might help, though AFAICT pretty much all the usize-s are line numbers here.

-------

In general, a few tests that illustrate how these objects behave and/or a few comments describing the algorithm would help.

Reply:
It's not that bad, but you could decrease the time it takes somebody to navigate the code, yes.

Reply:
I've added documentation to the functions. Does it seem more manageable now?

Reply:
The documentation is great, although it is written in a different tense/style than all other documentation. Since this is a minor thing, it could be improved in a follow-up.